src/lib.rs

```diff
@@ -14,46 +14,109 @@
+            "br" => Ok(Encoding::Brotli),
+            "identity" => Ok(Encoding::Identity),
+            "*" => Ok(Encoding::None),
+            _ => Err(ErrorKind::UnknownEncoding)?,
+        }
+    }
+
+    /// Converts the encoding into its' corresponding header value.
+    ///
+    /// Note that [`Encoding::None`] will return a HeaderValue with the content `*`.
+    /// This is likely not what you want if you are using this to generate the `Content-Encoding` header to be included in an encoded response.
+    pub fn to_header_value(self) -> HeaderValue {
+        match self {
+            Encoding::Gzip => HeaderValue::from_str("gzip").unwrap(),
+            Encoding::Deflate => HeaderValue::from_str("deflate").unwrap(),
+            Encoding::Brotli => HeaderValue::from_str("br").unwrap(),
+            Encoding::Identity => HeaderValue::from_str("identity").unwrap(),
+            Encoding::None => HeaderValue::from_str("*").unwrap(),
+        }
+    }
 }
 
 /// Parse a set of HTTP headers into an `Encoding`.
 pub fn parse(headers: &HeaderMap) -> Result<Encoding> {
-  let header = match headers.get(ACCEPT_ENCODING) {
-    Some(header) => header,
-    None => return Ok(Encoding::None),
-  };
-
-  let string = header.to_str().context(ErrorKind::InvalidEncoding)?;
-
-  if string.contains("br") {
-    Ok(Encoding::Brotli)
-  } else if string.contains("deflate") {
-    Ok(Encoding::Deflate)
-  } else if string.contains("gzip") {
-    Ok(Encoding::Gzip)
-  } else {
-    Err(ErrorKind::UnknownEncoding)?
-  }
+    let mut preferred_encoding = Encoding::None;
+    let mut max_qval = 0.0;
+
+    for header_value in headers.get_all(ACCEPT_ENCODING).iter() {
+        let header_value = header_value.to_str().context(ErrorKind::InvalidEncoding)?;
+        for v in header_value.split(',').map(str::trim) {
+            let mut v = v.splitn(2, ";q=");
+            let encoding = v.next().unwrap();
+
+            match Encoding::parse(encoding) {
+                Ok(encoding) => {
+                    if let Some(qval) = v.next() {
+                        let qval = match qval.parse::<f32>() {
+                            Ok(f) => f,
+                            Err(_) => return Err(ErrorKind::InvalidEncoding)?,
+                        };
+                        if (qval - 1.0f32).abs() < 0.01 {
+                            preferred_encoding = encoding;
+                            break;
+                        } else if qval > 1.0 {
+                            return Err(ErrorKind::InvalidEncoding)?; // q-values over 1 are unacceptable
+                        } else if qval > max_qval {
+                            preferred_encoding = encoding;
+                            max_qval = qval;
+                        }
+                    } else {
+                        preferred_encoding = encoding;
+                        break;
+                    }
+                }
+                Err(_) => continue, // ignore unknown encodings for now
```

Comment:
What's the failure mode we want here? What happens when we ignore this?

Reply:
So this part just helps gracefully ignore encodings that we don't know about -- IANA keeps a registry of all the possible encodings [here](https://www.iana.org/assignments/http-parameters/http-parameters.xml#http-parameters-1), and while I could just include all of the officially listed encodings, I opted for only making the most popular available.

If, say, the client sends a request with the header `Accept-Encoding: zstd;q=1.0, br;q=0.5`, instead of saying we don't understand the most preferred encoding that's being given to us, technically it *is* a valid header so we should at least return an encoding that we understand.

But if there's another preferred behavior for this you have in mind, then let me know :+1:

Reply:
Ahhhh, yep nope that makes perfect sense now! :100: